### DISCUSSION TITLE
Applying periodic boundary conditions on subdomain restricted variables

### DISCUSSION THREAD

**Discussion by ngrilli** [idaholab/moose#20023] [Q&A General] [answered]
Dear MOOSE Users,

As the title says, I would like to know if it is possible in MOOSE to apply
periodic boundary conditions on variables that are restricted to a subdomain/block.
Of course the boundaries on which I want to apply the periodic BC will be the
boundaries of the subdomain.

Does anybody have a successful example of this?

I try to explain what I tried:

I have a structured mesh and I divide it into
an "inner_volume" and "outer_volume".
In the inner_volume I use a model with advection,
representing dislocation transport with some variables that are restricted to the inner_volume.
I define the surfaces of the inner_volume using SubdomainBoundingBoxGenerator,
it is just a parallelepiped.
I use the "translation" option to define the periodic BC on the subdomain.
The outer_domain has a phenomenological model without transport.
Tensor mechanics kernels are applied everywhere,
while the dislocation variables have their own kernels restricted to inner_volume.
Two different materials calculate the stress in the two domains.

I show the RVE here: 

![immagine](https://user-images.githubusercontent.com/22821173/149160875-ea47bd64-f2d9-42b8-b358-062f1ea19e2c.png)

The results I obtain show that my advected quantity in the inner_volume just accumulates at the boundary,
indicating that Periodic BC are not applied and standard Neumann BC hold:

![immagine](https://user-images.githubusercontent.com/22821173/149160948-45f95518-fe08-41c4-9c2c-1874158dc7f6.png)

Moreover, the variable does not look periodic, so I think the Periodic BC is just skipped.
I do not get any warning or error though.

I have to mention the model with transport works very well
when I run the simulation with a single domain, with only the transport model.
In that case, I see the periodic boundary conditions applied properly.
Even though the model is complicated, I have tested it for months with periodic BC
on a single domain and I am sure it works well,
the problem only arises when I switch to two domains
and try to apply periodic BC on the subdomain.

I tried other things:
1) use "auto" periodic BC on the whole domain
2) define the surfaces in gmsh instead

but in no case I see the periodic BC applied properly on the subdomain

I attach the complete input file here.

Thank you very much in advance.
Nicolò

[FullGrain146_to_post.txt](https://github.com/idaholab/moose/files/7855344/FullGrain146_to_post.txt)

**Comment by ngrilli** [comment, marked as answer]
I want to add that I checked carefully in paraview that the surfaces I defined are correct and same for the translation vectors.
Is there any other thing I can try apart from Periodic BC, like some type of constraint or penalty terms between arbitrary nodes in the mesh?
Thank you
Nicolò

**Reply by jiangwen84** [reply]
This issue can be probably reproduced on a smaller and simpler problem, e.g, 2 elements and each one is a subdomain,  correct?

**Reply by ngrilli** [reply]
@jiangwen84 
Thank you for your answer, sure I will try to produce a minimal example.

Actually I noticed in the MOOSE tests this one:
https://github.com/idaholab/moose/blob/next/test/tests/bcs/periodic/periodic_subdomain_restricted_test.i

which seems to do the same, but I need to check if the result is actually correct.
I will keep you updated.

Nicolò

**Reply by GiudGiud** [reply]
@ngrilli any update on this?
